### PR TITLE
PyTorch Lightning Demo (old API)

### PR DESCRIPTION
**Pull request by romovpa**
## Opacus + Lightning

[PyTorch Lightning](https://www.pytorchlightning.ai/) is one of the most used frameworks for PyTorch. It helps to remove boilerplate code, improve readability and reproducibility. Lightning does this by restructuring the code (and not changing), uses core PyTorch objects without wrapping them.

We want:
1. Provide PL users with a way to use Opacus
2. Adapt our API to add privacy learning with minumin changes (ideally with no changes) to LightningModule

## Proposed Changes

**Example** `mnist_lightning.py`. 
Demonstration of how to add privacy-preserving learning to PL-based code. This is a clone of `mnist.py` written with PL. The code is self-descriptive and uses [best practices](https://pytorch-lightning.readthedocs.io/en/latest/starter/style_guide.html).

**Fixed** important bug in DPOptimizer causing #147. 
The fix makes the example workable. Optimizer closure was used incorrectly in `dp_step()`: the model is not reevaluated before making a DP-step, as the result we don't have grad_samples to make a step. I believe this can affect training not only in the Lightning scenario.

_This is for the main branch, not v1.0._ 
Merging this change with v1.0 is straightforward, but we need to disassemble `make_private()` into parts. The code won't be the most elegant, but this approach is transparent and easy to debug/undestand.

## New API and Full Integration with Lightning

Creating a seamless integration is a challenging task:
- Opacus requires dependencies between model, optimizer and dataloader. Whereas Lightning forces separation between them.
- From Ligthning's perspective, Opacus implements a custom [`pl.Trainer`](https://pytorch-lightning.readthedocs.io/en/latest/common/trainer.html) flow. `Trainer` is not designed to be extended through inheritance. 
  - A custom Trainer could be a straightforward solution for user, but it works only in short-term and very hard to maintain. 
- `Trainer` promotes using [callbacks](https://pytorch-lightning.readthedocs.io/en/latest/extensions/callbacks.html) and [custom loops](https://pytorch-lightning.readthedocs.io/en/latest/extensions/loops.html) for extension. This seems to me the only viable way to integrate. In this case `Trainer` will need to either 
  1. mutate modules (attach and detach privacy-related stuff)
  2. check and accept only valid modules, check that LightningModule and DataModule is properly tied, throw exception otherwise

I will follow up with some options.

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the **CONTRIBUTING** document and completed the CLA (see **CONTRIBUTING**).
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

**Comment by alexandresablayrolles** [comment]
LGTM, thanks @romovpa for the diff!

**Comment by romovpa** [comment]
[Implemented a prototype](https://github.com/romovpa/opacus/pull/5) of the better API-level integration with PL.

A PyTorch-Lightning user does:
```python
# Modeling as usual. No privacy!
class LitModel(LightningModule):
    def __init__(self):
        super().__init__()
        ...

    def forward(self):
        z = torch.rand(1, 3)
        image = self.decoder(z)
        image = image.view(1, 1, 28, 28)
        return image


def main():
    # Initialize model and data loader. No privacy here.
    model = LitModel()
    data = MNISTDataModule()

    # Adding some differential privacy: creating an engine, wrapping the DataModule
    privacy_engine = LightningPrivacyEngine(
        delta=1e-5,
        sample_rate=0.001,
        # etc
    )
    privacy_data = privacy_engine.wrap_datamodule(data)

    # Training
    trainer = pl.Trainer(
        max_epochs=2,
        enable_model_summary=False,
        callbacks=[privacy_engine],  # adding callback provided with the PrivacyEngine
    )
    trainer.fit(model, privacy_data)  # using a valid data module

    # This will crash with the message: 
    # Either remove PrivacyEngine callback or use certified data module
    trainer.fit(model, data)

    # Testing is as usual, no privacy here
    trainer.test(model, data)
    trainer.test(model, privacy_data)  # this also works

```

cc @ffuuugor @alexandresablayrolles

**Comment by romovpa** [comment]
Anyway, I suggest to merge & close this PR to guarantee that we have at least something for PL users, if they want to use Opacus. As a follow up I will try to suggest an API in `experimental_v1.0` branch.

To implement the idea above I will probably need to [redesign `PrivacyEngine` as a factory](https://github.com/pytorch/opacus/pull/226#discussion_r733789307), if there is no objection. This approach divides hyperparameters (settings) and the arguments to which DP is applied (model, data, opt). I need this separation for a clean integration with PL.

**Comment by ffuuugor** [comment]
Thanks!
Agree, let's merge it to main ASAP - I see no reason not to.
If you import the diff, we'll accept 

re: factory redesign 
I was going to implement the proposal almost exactly like stated in the comment you linked.
One difference - because of unlucky timing of PyTorch DevDay recordings, we've already committed to a certain API, so we probably want to also support simple constructor object creation too.
In terms of which parameters go where, I feel strongly (and also we've committed to in the presentation) that noise_multiplier and max_grad_norm belong in `make_private`.
`poisson_sampling` should probably be moved to constructor (and therefore factory), as it defines the behaviour strategy, not just parameters. 

I was going to implement that over the next couple days (certainly before Tuesday) - does this timeline work for you? If not, feel free to do the draft implementation and we can converge later

**Review by ffuuugor — APPROVED** [review]

File: opacus/privacy_engine.py [hunk 1]
```diff
@@ -350,15 +350,22 @@ def dp_step(self, closure=None, is_empty=False):
...
                 # We just update the accountant
                 self.privacy_engine.steps += 1
-
+                return self.original_step(closure)
             else:
+                loss = None
+                if closure is not None:
+                    with torch.enable_grad():
+                        loss = closure()
+
```
Comment (karthikprasad) [inline opacus/privacy_engine.py:359]: noob: Why isn't this necessary when module has `ddp_hooks`? Why can't this be common for both if and else blocks?

Comment (romovpa) [reply opacus/privacy_engine.py:359]: Good question. It is worth explaining. 

First, wth is the closure for. From [PyTorch docs](https://pytorch.org/docs/stable/optim.html):

> Some optimization algorithms such as Conjugate Gradient and LBFGS need to reevaluate the function multiple times, so you have to pass in a closure that allows them to recompute your model. The closure should clear the gradients, compute the loss, and return it.
> 
> Example:
> ```python
>     def closure():
>         optimizer.zero_grad()
>         output = model(input)
>         loss = loss_fn(output, target)
>         loss.backward()
>         return loss
> ```

Before clipping and adding noise in `self.privacy_engine.step(is_empty)` we need to compute the model. Otherwise there is nothing to apply clipping/noise to. 

In case of `ddp_hooks` we just call `self.original_step(closure)` and pass the closure to call it at the original oprimiser's discretion. As the comment suggests, the clipping and noising are performed by the hooks at the end of the backward pass.

File: opacus/privacy_engine.py [hunk 2]
```diff
@@ -350,15 +350,22 @@ def dp_step(self, closure=None, is_empty=False):
...
                 self.privacy_engine.step(is_empty)
                 if isinstance(
                     self.privacy_engine.module._module,
                     DifferentiallyPrivateDistributedDataParallel,
                 ):
                     average_gradients(self.privacy_engine.module)
-            self.original_step(closure)
+
+                self.original_step()
```
Comment (romovpa) [inline opacus/privacy_engine.py:367]: Hmm, I'm not sure if I should pass the closure to the original_step. 
- If I do, it probably calls it again and we run the backward pass twice. 
- If I do NOT, the optimizers requiring several model reevaluations must not work. Although don't know how that case affects grad sampling and accounting.

@alexandresablayrolles maybe you know the answer?

File: opacus/privacy_engine.py (hunk 1 above)
Comment (karthikprasad) [reply opacus/privacy_engine.py:359]: I see. Thanks @romovpa . Can you perhaps add the above explanation as a comment to the code. (Unless, it is obvious for everyone.)
